src/payload.iced

```diff
@@ -0,0 +1,45 @@
+{prng} = require('crypto')
+{createHash} = require('crypto')
+{createHmac} = require('crypto')
+msgpack = require('purepack')
+crypto = require('keybase-nacl')
+nonce = require('./nonce.iced')
+util = require('./util.iced')
+
+# HMAC-SHA512-256(step1 hash, mac_key)
+compute_authenticator = (hash, key) ->
+  hmac = createHmac('sha512', key)
+  hmac.update(hash)
+  return hmac.digest()[0...32]
+
+# sha512(header_hash + payload secretbox nonce + payload secretbox)
+step1 = (header_hash, block_num, payload_secretbox) ->
+  step1_cat = Buffer.concat([header_hash, nonce.nonceForChunkSecretBox(block_num), payload_secretbox])
+  crypto_hash = createHash('sha512')
+  crypto_hash.update(step1_cat)
+  step1_hash = crypto_hash.digest()
+  return step1_hash
+
+exports.generate_encryption_payload_packet = (payload_encryptor, plaintext, block_num, header_hash, mac_keys) ->
+  # perform nacl encryption of payload
+  payload_secretbox = payload_encryptor.secretbox({plaintext, nonce : nonce.nonceForChunkSecretBox(block_num)})
+
+  # compute the authenticators
+  step1_hash = step1(header_hash, block_num, payload_secretbox)
+  authenticators = []
+  for mac_key in mac_keys
+    authenticator = compute_authenticator(step1_hash, mac_key)
+    authenticators.push(authenticator)
+
+  return [authenticators, payload_secretbox]
+
+exports.parse_encryption_payload_packet = (payload_decryptor, payload_list, block_num, header_hash, mac_key, recipient_index) ->
+  # verify that we are an authenticator
+  step1_hash = step1(header_hash, block_num, payload_list[1])
+  computed_authenticator = compute_authenticator(step1_hash, mac_key)
+  unless util.bufeq_secure(computed_authenticator, payload_list[0][recipient_index]) then throw new Error('Integrity check failed!')
```

Comment:
Just thinking out loud: This is one of the reasons I don't like that `buffeq_secure` returns true when both of its arguments are `null`/`undefined`. The expression `payload_list[0][recipient_index]` will be `undefined` if an attacker gives us a short authenticators list. That means that if `computed_authenticator` ever returned `null`, an attacker could bypass authentication here. Of course, `computed_authenticator` doesn't return null, and it probably never will. But that's the kind of thing someone might not realize if they're rewriting that function in 5 years.

Anyway, no need to change it. But if we decide to rewrite `buffeq_secure` later maybe we can make this change :)